torch/distributed/fsdp/_symbolic_trace.py

```diff
@@ -0,0 +1,220 @@
+
+    Args:
+        tracer (torch.fx.Tracer): An instance of ``torch.fx.Tracer`` that will be used to perform symbolic
+            tracing. ``tracer`` is default to be ``torch.fx.Tracer()``, but can also be instance
+            of some child class of ``torch.fx.Tracer``. For example, one may want to use ``HFTracer``
+            for models in Transformers:
+            .. _Transformers: https://huggingface.co/docs/transformers/index
+
+        concrete_args (Optional[Dict[str, Any]]): Concrete arguments that should not be treated
+            as ``torch.fx.Proxy`` when tracing the forward function.
+            ``concrete_args`` allows one to partially specialize the forward function,
+            including removing control flow or data structures.
+            ``concrete_args`` is also the parameter used in :meth:`~torch.fx.Tracer.trace`.
+    """
+
+    tracer: torch.fx.Tracer = torch.fx.Tracer()
+    concrete_args: Optional[Dict[str, Any]] = None
+
+
+@dataclass
+class _ExecutionInfo:
+    """
+    Contains the execution order information in the model forward pass.
+
+    Attributes:
+        current_module: record the module that is currently being traced.
+
+        module_forward_order: a list of modules, where the ordering is based on
+            when their forward function is called.
+
+        param_exec_order: a list of parameters ordered based on their execution order.
+
+        _traced_param_set: a set containing all parameters that have been traced.
+
+        module_execution_info_dict: a dict that maps each module to a list of
+            tuple containing a module and a list of named parameters.
+            For a given module, each tuple:
+            1. either contains this module and part of its ``named_parameters`` that will be executed together,
+            2. or contains one of its child modules and all of the child module's ``named_parameters``.
+            The list of tuple is ordered based on the parameter execution order.
+    """
+
+    current_module: torch.nn.Module
+    module_forward_order: List[torch.nn.Module]
+    module_execution_info_dict: Dict[
+        torch.nn.Module,
+        List[Tuple[torch.nn.Module, List[Tuple[str, torch.nn.Parameter]]]],
+    ]
+    param_exec_order: List[torch.nn.Parameter] = field(default_factory=list)
+    _traced_param_set: Set[torch.nn.Parameter] = field(default_factory=set)
+
+
+def _init_execution_info(root_module: torch.nn.Module) -> _ExecutionInfo:
+    """
+    Create an instace of _ExecutionInfo with initialization based on ``root_module``.
+
+    Args:
+        root_module (torch.nn.Module): the module to get the execution information
+            via ``tracer.trace()`` inside ``_patch_tracer``.
+    """
+    return _ExecutionInfo(
+        current_module=root_module,
+        module_forward_order=[root_module],
+        module_execution_info_dict={root_module: []},
+    )
+
+
+def _patched_create_proxy(
```

Comment:
Is this patching approach considered best practice when working with FX? It seems that a lot of this can be done with FX providing pre and post hooks to create_proxy, call_module, etc - but not sure if FX has those hooks in place at the moment.

Reply:
Agreed. We should be sure that patching is the preferred solution here.

Reply:
@rohan-varma I agree that pre and post hooks would work the best, but I don't think currently they are supported in FX. I also have discussed with Shen about optional ways to implement this: https://github.com/pytorch/pytorch/pull/80294#discussion_r909170775
but I haven't found any way that could implement this cleanly without patching.